### PR TITLE
W-14559071: perform AST validation before macro expansion

### PR DESCRIPTION
**Pull request by hyerans**

### CODE REVIEW

**Review by fsgonz — COMMENTED** [review]
We need a test for this.

**Review by fsgonz — CHANGES_REQUESTED** [review]
we need a test for this.

File: modules/spring-config/src/main/java/org/mule/runtime/config/internal/context/lazy/LazyMuleArtifactContext.java [hunk 1]
```diff
@@ -175,7 +175,6 @@ public LazyMuleArtifactContext(MuleContext muleContext, ArtifactAst artifactAst,
     this.artifactProperties = artifactProperties;
     this.runtimeLockFactory = runtimeLockFactory;
 
-    initialize();
```
Comment (fsgonz) [inline modules/spring-config/src/main/java/org/mule/runtime/config/internal/context/lazy/LazyMuleArtifactContext.java:178]: What has to be done before the initialize is the ast validation. But according to the comment below the graph should be done after the initialize. Do we have a way to guaranteee that we validate the ast without macroexpansion?

**Comment by hyerans** [comment]
--validate

File: modules/spring-config/src/main/java/module-info.java [hunk 2]
```diff
@@ -82,6 +82,7 @@
   // Spring JNDI support
   requires java.naming;
   requires java.transaction;
+  requires org.mule.runtime.artifact.ast;
```
Comment (elrodro83) [inline modules/spring-config/src/main/java/module-info.java:85]: group mule dependencies at the beginning of the file, alphabetically

File: modules/spring-config/src/main/java/org/mule/runtime/config/internal/context/lazy/ComponentInitializationRequest.java [hunk 3]
```diff
@@ -69,11 +70,13 @@ public static class Builder {
      * @param keepPrevious                    Whether previously initialized components should be kept unchanged.
      */
     public Builder(ArtifactAstDependencyGraph graph,
+                   ArtifactAstDependencyGraph graphWithoutMacroExpansion,
```
Comment (elrodro83) [inline modules/spring-config/src/main/java/org/mule/runtime/config/internal/context/lazy/ComponentInitializationRequest.java:73]: is the macroexpanded graph required for anything?

**Review by elrodro83 — CHANGES_REQUESTED** [review]

File: modules/spring-config/src/main/java/org/mule/runtime/config/internal/context/lazy/ComponentInitializationRequest.java (hunk 3 above)
Comment (hyerans) [reply modules/spring-config/src/main/java/org/mule/runtime/config/internal/context/lazy/ComponentInitializationRequest.java:73]: Graph without macro expansion was added for the validation with lazy init. 
_Graph_(the macroexpanded graph) has been used/ and is still being used when we get minimalAst to update the requestedLocations during the creation of components.

**Comment by hyerans** [comment]
--validate

File: modules/spring-config/src/main/java/org/mule/runtime/config/internal/context/lazy/ComponentInitializationRequest.java [hunk 4]
```diff
@@ -187,7 +201,15 @@ public Set<String> getRequestedLocations() {
    * @throws ConfigurationException If the validation fails.
    */
   public void validateRequestedAst(ArtifactAstValidator astValidator) throws ConfigurationException {
-    astValidator.validate(getMinimalAst());
+    astValidator.validate(getMinimalAstWithoutMacroExpansion());
+  }
+
+  private ArtifactAst getMinimalAstWithoutMacroExpansion() {
```
Comment (elrodro83) [inline modules/spring-config/src/main/java/org/mule/runtime/config/internal/context/lazy/ComponentInitializationRequest.java:207]: rename to `getBaseMinimalArtifactAst()`

File: modules/spring-config/src/main/java/org/mule/runtime/config/internal/context/lazy/ComponentInitializationRequest.java [hunk 5]
```diff
@@ -221,15 +243,15 @@ private static Predicate<ComponentAst> buildFilterFromLocation(Location location
...
-    if (minimalArtifactAst == null) {
-      minimalArtifactAst = doGetMinimalAst();
+    if (postProcessedMinimalArtifactAst == null) {
+      postProcessedMinimalArtifactAst = doGetMinimalAst();
     }
-    return minimalArtifactAst;
+    return postProcessedMinimalArtifactAst;
   }
 
   private ArtifactAst doGetMinimalAst() {
```
Comment (elrodro83) [inline modules/spring-config/src/main/java/org/mule/runtime/config/internal/context/lazy/ComponentInitializationRequest.java:252]: rename to `getPostProcessedMinimalArtifactAst()`

**Review by elrodro83 — APPROVED** [review]

**Review by elrodro83 — APPROVED** [review]

**Pull request merged**